adagrams/game.py

```diff
@@ -1,11 +1,171 @@
+
+    # return list
+    return hand
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+    # create variables to hold data structures
+    letter_list = list(word.upper())
+    letter_bank_freq = {}
+    letter_list_freq = {}
+
+    # loop through letter bank to count frequencies
+    for letter in letter_bank:
+        if letter in letter_bank_freq:
+            letter_bank_freq[letter] += 1
+        else:
+            letter_bank_freq[letter] = 1
+
+    # loop through letter list to check if letter is in bank and count frequences
+    for letter in letter_list:
+        if letter not in letter_bank:
+            return False
+        elif letter in letter_list_freq:
+            letter_list_freq[letter] += 1
+        elif not letter in letter_list_freq:
+            letter_list_freq[letter] = 1
+
+    # Check if letter is used more than avaiable times than in letter bank
+    if letter_list_freq[letter] > letter_bank_freq[letter]:
+        return False
+
+    # Return True if inputted word is valid given letter bank
+    return True
 
 def score_word(word):
-    pass
+    # assign int to variable for the score
+    score = 0
+    letter_list = list(word.upper())
+
+    # create a loop to add values of each letter to score
+    for letter in letter_list:
+        score += letter_values[letter]
+
+    # check length of word for bonus point 
+    if len(letter_list) >= 7:
+        score += 8
+
+    # return score
+    return score
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    # create dict to hold all word inputted by player
+    all_words_dict = {}
+    
+    # loop to add items to all words dict with word as key and score as value
+    for word in word_list:
+        score = score_word(word)
+        all_words_dict[word] = score
+    
+    # create a list to hold scores and sort scores
+    scores = all_words_dict.values()
+    score_list = list(scores)
+    score_list.sort()
+
+    # create list to hold best words that have the highest scores
+    best_word_list = []
+
+    # create loop to append highest scoring words to best word lists
+    for word in all_words_dict:
+        if all_words_dict[word] == score_list[-1]:
+            best_word_list.append(word)
+    
+    # sort best word list and list to hold best word lengths for tie comparisons
+    best_word_list.sort()
+    best_word_lengths = []
+
+    # loop through best word list to get lengths and add to best word lengths
+    for word in best_word_list:
+        best_word_lengths.append(len(word))
+
+    # sort best word lenghs
+    best_word_lengths.sort()
+
+    # conditionals to check for tie situations
+    for word in best_word_list:
+        if len(word) >= 10:
+            return best_word_list[0], score_list[-1]
+        elif best_word_lengths[0] == len(word):
+            return word, score_list[-1]
+
+    # return best word and score as tuple if no tie situations 
+    return best_word_list[-1], score_list[-1]
```

Comment:
This all works, but it definitely starts to get bulky again. While there's nothing inherently wrong with separating it out into a list equating each word with its score and then sorting (Great use of the sort method btw), we can drastically cut down on the code in this function by doing the comparisons in place! What this means is creating some variable (tuple, a string and an int, etc.) to hold the first word in the list and its score. From there we go through the rest of the list and make the comparisons in a single for loop to figure out if a word should replace the variable/s we have to hold the highest scoring word and score! This is just a matter of figuring out the correct conditions for replacement but it does allow for a single loop to control our selection!